### PR TITLE
GEOMESA-642 Implement raster files ingestion using Scalding

### PR DESCRIPTION
**Pull request by wjp22903**

### CODE REVIEW

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/index/IndexEntry.scala [hunk 1]
```diff
@@ -142,4 +142,16 @@ trait IndexHelpers {
...
+case class DecodedIndex(id: String, geom: Geometry, dtgMillis: Option[Long]) {
+  def toBytes(): Array[Byte] =  {
+    val encodedId = id.getBytes
+    val encodedGeom = WKBUtils.write(geom)
+    val date = dtgMillis.map{new DateTime(_, DateTimeZone.UTC)}
+    val encodedDtg = date.map { dtg => ByteBuffer.allocate(8).putLong(dtg.getMillis).array() }
+                         .getOrElse(Array[Byte]())
+
+    ByteBuffer.allocate(4).putInt(encodedId.length).array() ++ encodedId ++
+      ByteBuffer.allocate(4).putInt(encodedGeom.length).array() ++ encodedGeom ++
```
Comment (jnh5y) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/index/IndexEntry.scala]: Is there a way to allocate one ByteBuffer of the correct size and call .array once?  If so, I think that will be more efficient.

File: geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/data/Raster.scala [hunk 2]
```diff
@@ -38,9 +39,31 @@ case class Raster(chunk: RenderedImage, metadata: DecodedIndex, resolution: Doub
...
 
+  def encodeValue = RasterUtils.imageSerialize(chunk)
+
+  def encodeToBytes(): Array[Byte] = {
+    val chunkBytes = RasterUtils.imageSerialize(chunk)
+    val metaDataBytes = metadata.toBytes
+
+    ByteBuffer.allocate(4).putInt(chunkBytes.length).array() ++ chunkBytes ++
+      ByteBuffer.allocate(4).putInt(metaDataBytes.length).array() ++ metaDataBytes ++
+      ByteBuffer.allocate(8).putDouble(resolution).array()
```
Comment (jnh5y) [inline geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/data/Raster.scala]: Same comment as above.  During a later ticket, we may want to explore Avro encoding for this info.

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/index/IndexEntry.scala (hunk 1 above)
Comment (wjp22903) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/index/IndexEntry.scala]: Good point. I borrowed it from some other places in Geomesa. Wii do.

File: geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/data/Raster.scala [hunk 3]
```diff
@@ -38,9 +39,31 @@ case class Raster(chunk: RenderedImage, metadata: DecodedIndex, resolution: Doub
...
     s"${rasterName}_${UUID.randomUUID.toString}"
+
+  def apply(bytes: Array[Byte]): Raster = {
+    val chunkLength = ByteBuffer.wrap(bytes, 0, 4).getInt
+    val (chunkPortion, metaResPortion) = bytes.drop(4).splitAt(chunkLength)
+    val chunk = RasterUtils.imageDeserialize(chunkPortion)
+    val metaLength = ByteBuffer.wrap(metaResPortion, 0, 4).getInt
+    val (metaPortion, resPortion) = metaResPortion.drop(4).splitAt(metaLength)
+    val metaData = IndexEntry.byteArrayToDecodedIndex(metaPortion)
+    val resolution = ByteBuffer.wrap(resPortion).getDouble
```
Comment (jnh5y) [inline geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/data/Raster.scala]: Similar to above, I'd be interested in thinking through this block's use of ByteBuffer.  The calls to drop and splitAt use implicits.  I'm unsure of the impact that has on speed and memory usage. 

If ByteBuffer's get(byte[] dst, int offset, int length) makes this cleaner, then feel free.  If that's not fast, let's add a ticket for reviewing this block and move on.

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala [hunk 4]
```diff
@@ -38,75 +35,104 @@ class IngestRasterCommand(parent: JCommander) extends Command with AccumuloPrope
...
+          case Success(info) => logger.info("Local ingestion is done.")
+          case Failure(e) => throw new RuntimeException(e)
+        }
+      case "remote" =>
+        val rasterSerializer =
+          new RasterFilesSerialization(baseRasterIngestParams + (IngestRasterParams.FILE_PATH  -> Some(params.file)))
+        rasterSerializer.runSerializationTask() match {
+          case Success(outPath) =>
+            logger.info("Raster files serialization is done.")
+            new RemoteRasterIngest(baseRasterIngestParams + (IngestRasterParams.FILE_PATH  -> Some(outPath))).run
```
Comment (jnh5y) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala]: Does running the RemoteRasterIngest job need any more logging here?

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala [hunk 5]
```diff
@@ -38,75 +35,104 @@ class IngestRasterCommand(parent: JCommander) extends Command with AccumuloPrope
...
+          case Failure(e) => throw new RuntimeException(e)
+        }
+      case _ => logger.error("Unsupported ingestion mode")
```
Comment (jnh5y) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala]: If a user mistypes something, can we hit this error?  If so, I'd suggest expanding it to hint at usage.

File: geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala [hunk 6]
```diff
@@ -38,75 +35,104 @@ class IngestRasterCommand(parent: JCommander) extends Command with AccumuloPrope
...
     }
   }
 
-  def createCoverageStore(config: IngestRasterParameters): AccumuloCoverageStore = {
-    val password = getPassword(params.password)
-    val csConfig: JMap[String, Serializable] = getAccumuloCoverageStoreConf(config, password)
+  def getFormat(fileOrDir: File): String = {
+    val file =
+      if (fileOrDir.isDirectory) fileOrDir.listFiles.head
```
Comment (jnh5y) [inline geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/commands/IngestRasterCommand.scala]: Do we provide a useful error if the user can a mismatch of input files?

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/index/IndexEntry.scala (hunk 1 above)
Comment (aannex-ccri) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/index/IndexEntry.scala]: this looks like repeated code, I am almost positive this function exists in core already.

**Comment by jnh5y** [comment]
Merged on f_wcs_4 as 0494c5a4c9962c574dfbcaf959e9206135155cf5.

**Pull request closed without merging**